abi.md

```diff
@@ -110,6 +110,178 @@ Passive bundles shall be lowered to Verilog packed structs.
+The bind lowering convention is indicated by the `"bind"`{.firrtl} string on a
+declared group.  When using this convention, optional groups are lowered to
+separate modules that are instantiated via SystemVerilog `bind`{.verilog}
+statements ("bind instantiation").
+
+Each module that contains group instances will produce one additional module per
+group that has "internal" module convention.  I.e., the modules are private and
+have no defined ABI---the name of each module is implementation defined, the
+instantiation name of a bound instance is implementation defined, and any ports
+created on the module may have any name and may be optimized away.
+
+Practically speaking, additional ports of each generated module must be created
+whenever a value defined outside the group is used by the group.  The values
+captured by the group will create input ports.  Values defined in a group and
+used by a nested group will create output ports because SystemVerilog disallows
+the use of a bind instantiation underneath the scope of another bind
+instantiation.  The names of additional ports are implementation defined.
+
+For each optional group, one binding file will be produced for each group or
+nested group.  The circuit, group name, and any nested groups are used to derive
+a predictable filename of the binding file.  The file format uses the format
+below:
+
+``` ebnf
+filename = "groups_" , circuit , "_", root , { "_" , nested } , ".sv" ;
+```
+
+As an example, consider the following circuit with three optional groups:
+
+``` firrtl
+circuit Foo:
+  declgroup "bind", Group1:
+    declgroup "bind", Group2:
+      declgroup "bind", Group3:
+```
+
+When compiled to Verilog, this will produce three bind files:
+
+```
+groups_Foo_Group1.sv
+groups_Foo_Group1_Group2.sv
+groups_Foo_Group1_Group2_Group3_.sv
+```
+
+The contents of each binding files must have the effect of including all code
+defined in a group or its parent groups.
+
+#### Example
+
+The end-to-end below example shows a circuit with one two groups that is lowered
+to Verilog.  This shows example Verilog output which is not part of the ABI, but
+is included for informative reasons.
+
+Consider the following circuit containing one optional group, `Group1`, and one
+nested optional group, `Group2`.  Module `Foo` contains one instantiation of
+module `Bar`.  Both `Foo` and `Bar` contain groups.  To make the example
+simpler, no constant propagation is done:
+
+``` firrtl
+circuit Foo:
+  declgroup "bind", Group1:
+    declgroup "bind", Group2:
+
+  module Bar:
+    wire a: UInt<1>
+    connect a, UInt<1>(0)
+
+    group Group1:
+      output _notA: Probe<UInt<1>>
+      node notA = not(a)
+      define _notA = probe(notA)
+
+      group Group2:
+        output _notNotA: Probe<UInt<1>>
+        node notNotA = not(notA)
+        define _notNotA = probe(notNotA)
+
+  module Foo:
+    inst bar of Bar
+
+    group Group1:
+      node x = bar.notA
+
+      group Group2:
+        node y = bar.notNotA
```

Comment:
```suggestion
  module Foo:
    inst bar of Bar

    group Group1:
      node x = bar._notA

      group Group2:
        node y = bar._notNotA
```

I think this is currently referencing the wrong signals? Also I think the Verilog has the same inconsistency.